### PR TITLE
@Laravel rules

### PR DESCRIPTION
**Pull request by valentinxxx**
Based on [Style CI Laravel preset](https://styleci.readme.io/v1.0/docs/presets#section-laravel).
Assuming that `@PSR2` rules represents the StyleCI's [PSR2 preset](https://styleci.readme.io/v1.0/docs/presets#section-psr2), I've made a diff:

- `+align_phpdoc`<br>
refactored `PhpdocAlignFixer` with `parts` and `separatorSpaces` options<br>
`parts`: `hint`, `var`, `desc`<br>
- `+binary_operator_spaces`<br>
`+unalign_equals`<br>
added:
```
'binary_operator_spaces' => [
    'align_double_arrow' => null,
    'align_equals' => false,
]
```
- `+blank_line_after_opening_tag`<br>
added: `'blank_line_after_opening_tag' => true`
- `+blank_line_before_return`<br>
added:
```
'blank_line_before_statement' => [
    'statements' => ['return'],
],
```
- `+cast_spaces`<br>
added: `'cast_spaces' => true`
- `+concat_without_spaces`<br>
added: `'concat_space' => ['spacing' => 'none']`
- `+declare_equal_normalize`<br>
added: `'declare_equal_normalize' => true`
- `+function_typehint_space`<br>
added: `'function_typehint_space' => true`
- `+hash_to_slash_comment`<br>
added: `'hash_to_slash_comment' => true`
- `+heredoc_to_nowdoc`<br>
added: `'heredoc_to_nowdoc' => true`
- `+include`<br>
added: `'include' => true`
- `+length_ordered_imports`<br>
added: `'ordered_imports' => ['sortAlgorithm' => 'length']`
- `+lowercase_cast`<br>
added: `'lowercase_cast' => true`
- `+magic_constant_casing`<br>
added: `'magic_constant_casing' => true`
- `+method_separation`<br>
added: `'method_separation' => true`
- `+native_function_casing`<br>
added: `'native_function_casing' => true`
- `+no_alias_functions`<br>
added: `'no_alias_functions' => true`
- `+no_blank_lines_after_class_opening`<br>
added: `'no_blank_lines_after_class_opening' => true`
- `+no_blank_lines_after_phpdoc`<br>
added: `'no_blank_lines_after_phpdoc' => true`
- `+no_blank_lines_after_throw`<br>
TODO
- `+no_blank_lines_between_imports`<br>
`+no_blank_lines_between_traits`<br>
`+no_extra_consecutive_blank_lines`<br>
added:
```
'no_extra_consecutive_blank_lines' => ['tokens' => [
    'extra',
    'use',
    'use_trait',
]]
```
- `+no_empty_phpdoc`<br>
added: `'no_empty_phpdoc' => true`
- `+no_empty_statement`<br>
added: `'no_empty_statement' => true`
- `+no_leading_import_slash`<br>
added: `'no_leading_import_slash' => true`
- `+no_leading_namespace_whitespace`<br>
added: `'no_leading_namespace_whitespace' => true`
- `+no_multiline_whitespace_around_double_arrow`<br>
added: `'no_multiline_whitespace_around_double_arrow' => true`
- `+no_multiline_whitespace_before_semicolons`<br>
added: `'no_multiline_whitespace_before_semicolons' => true`
- `+no_short_bool_cast`<br>
added: `'no_short_bool_cast' => true`
- `+no_singleline_whitespace_before_semicolons`
added: `'no_singleline_whitespace_before_semicolons' => true`
- `+no_spaces_inside_offset`<br>
added: `'no_spaces_around_offset' => ['positions' => ['inside']]`
- `+no_trailing_comma_in_list_call`<br>
added: `'no_trailing_comma_in_list_call' => true`
- `+no_trailing_comma_in_singleline_array`
added: `'no_trailing_comma_in_singleline_array' => true`
- `+no_unneeded_control_parentheses`<br>
added: `'no_unneeded_control_parentheses' => true`
- `+no_unreachable_default_argument_value`<br>
added: `'no_unreachable_default_argument_value' => true`
- `+no_unused_imports`<br>
added: `'no_unused_imports' => true`
- `+no_useless_return`<br>
added: `'no_useless_return' => true`
- `+no_whitespace_before_comma_in_array`<br>
added: `'no_whitespace_before_comma_in_array' => true`
- `+no_whitespace_in_blank_line`<br>
added: `'no_whitespace_in_blank_line' => true`
- `+normalize_index_brace`<br>
added: `'normalize_index_brace' => true`
- `+not_operator_with_successor_space`<br>
added: `'not_operator_with_successor_space' => true`
- `+object_operator_without_whitespace`<br>
added: `'object_operator_without_whitespace' => true`
- `+phpdoc_indent`<br>
added: `'phpdoc_indent' => true`
- `+phpdoc_inline_tag`<br>
added: `'phpdoc_inline_tag' => true`
- `+phpdoc_no_access`<br>
added: `'phpdoc_no_access' => true`
- `+phpdoc_no_package`<br>
added: `'phpdoc_no_package' => true`
- `+phpdoc_no_useless_inheritdoc`<br>
added: `'phpdoc_no_useless_inheritdoc' => true`
- `+phpdoc_scalar`<br>
added: `'phpdoc_scalar' => true`
- `+phpdoc_single_line_var_spacing`<br>
added: `'phpdoc_single_line_var_spacing' => true`
- `+phpdoc_summary`<br>
added: `'phpdoc_summary' => true`
- `+phpdoc_to_comment`<br>
added: `'phpdoc_to_comment' => true`
- `+phpdoc_trim`<br>
added: `'phpdoc_trim' => true`
- `+phpdoc_type_to_var`<br>
added:
```
'phpdoc_no_alias_tag' => ['replacements' => [
    'type' => 'var',
]],
```
- `+phpdoc_types`<br>
added: `'phpdoc_types' => true`
- `+phpdoc_var_without_name`<br>
added: `'phpdoc_var_without_name' => true`
- `+print_to_echo`<br>
added: `'no_mixed_echo_print' => true`
- `+self_accessor`<br>
added: `'self_accessor' => true`
- `+short_array_syntax`<br>
added: `'array_syntax' => ['syntax' => 'short']`
- `+short_scalar_cast`<br>
added: `'short_scalar_cast' => true`
- `+simplified_null_return`<br>
added: `'simplified_null_return' => true`
- `+single_blank_line_before_namespace`<br>
added: `'single_blank_line_before_namespace' => true`
- `+single_quote`<br>
added: `'single_quote' => true`
- `+space_after_semicolon`<br>
added: `'space_after_semicolon' => true`
- `+standardize_not_equals`<br>
added: `'standardize_not_equals' => true`
- `+ternary_operator_spaces`<br>
added: `'ternary_operator_spaces' => true`
- `+trailing_comma_in_multiline_array`<br>
added: `'trailing_comma_in_multiline_array' => true`
- `+trim_array_spaces`<br>
added: `'trim_array_spaces' => true`
- `+unary_operator_spaces`<br>
added: `'unary_operator_spaces' => true`
- `+whitespace_after_comma_in_array`<br>
added: `'whitespace_after_comma_in_array' => true`

Additional rules:
- `method_argument_space`<br>
in `@PSR2` `method_argument_space`'s `ensure_fully_multiline` property is `true` -> override with default `false`<br>
added: `'method_argument_space' => true`

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Who will keep this up to date? I won't guarentee that when Laravel makes a change, that PHP-CS-Fixer will be updated by me. StyleCI is currently the authority on the definition of Laravel's code style.

File: README.rst [hunk 1]
```diff
@@ -933,32 +933,36 @@ Choose from the list of available rules:
...
     untyped parameters only; defaults to ``true``
 
-* **phpdoc_align** [@Symfony]
+* **phpdoc_align** [@Symfony, @Laravel]
 
   All items of the given phpdoc tags must be aligned vertically.
 
   Configuration options:
 
+  - ``parts`` (``array``): parts should be aligned; defaults to ``['hint', 'var',
```
Comment (GrahamCampbell) [inline README.rst]: Where did this come from?

**Comment by valentinxxx** [comment]
@GrahamCampbell Because of the .php_cs file is removed from Laravel (https://github.com/laravel/framework/pull/20001), it's only a nice to have for Laravel based projects which are not using StyleCI.
I think the community will update if needed. ;)

File: README.rst (hunk 1 above)
Comment (valentinxxx) [inline README.rst]: From the original regex: https://github.com/valentinxxx/PHP-CS-Fixer/blob/2.4/src/Fixer/Phpdoc/PhpdocAlignFixer.php#L63-L71

`<tag>`, `<hint>`, `<var>`, `<desc>`

**Comment by GrahamCampbell** [comment]
That original .php_cs framework file was never correct actually, anyway. :P

**Comment by keradus** [comment]
If nobody from Laravel core team will maintain the ruleset, it will become dead and I'm :-1: on it

File: src/RuleSet.php [hunk 2]
```diff
@@ -155,6 +155,46 @@
             'psr4' => true,
             'silenced_deprecation_error' => true,
         ],
+        '@Laravel' => [
+            '@Symfony' => true,
```
Comment (keradus) [inline src/RuleSet.php]: this is tricky unless `Laravel` decide to take `Symfony` CS guidelines with few additions. What I see, it is taking `PSR2` with few additions

**Comment by keradus** [comment]
that's what happen when you build one tool as wrapper of other tool and claim to support only wrapper.

**Comment by keradus** [comment]
If nobody from Laravel core team will maintain the ruleset, it will become dead and I'm :-1: on it

**Comment by barryvdh** [comment]
Imho, this styleset should only be used when contributing to Laravel directly. Own projects should use psr2. 
As stated in the docs, you also don't have to adhere to the style in pr's, because it will be fixed automatically. https://laravel.com/docs/5.4/contributions#coding-style

**Comment by GrahamCampbell** [comment]
@barryvdh I agree tbh. There's no reason for people to adopt the laravel core code style, just because they have a laravel project.

File: src/RuleSet.php (hunk 2 above)
Comment (valentinxxx) [inline src/RuleSet.php]: @keradus you are right, refactored ruleset, now based on `@PSR2`

**Comment by SpacePossum** [comment]
I think we would be nice if someone from the Laravel community would create a public repo with a default configuration file matching the rules as used within the Laravel project. This way people can reuse that in their own projects while the configuration is maintained by the Laravel community. 

The last part is important as it would be better fit than maintaining it here. Updates to the Laravel configuration would be bound to the releases of the fixer while such dependency is not needed at all.

Secondly when we update the fixer we would've to check if this would require an update of the Laravel rule set. We tried this for a long time for the Symfony set and its not ideal as releases might be deployed when waiting on feedback of the community or if we updated the set not to the liking of the community we've to update and release the fixer again.

For this and after discussion I'm closing this ticket. Feel free to further discuss this topic here.

**Pull request closed without merging**